### PR TITLE
Fix 17699 - Importing a module that has both modulename.d and modulename/package.d should be an error

### PR DESCRIPTION
**Pull request by dkorpel**

### CODE REVIEW

File: compiler/test/fail_compilation/test17699.d [hunk 1]
```diff
@@ -0,0 +1,12 @@
+/*
```
Comment (ibuclaw) [inline compiler/test/fail_compilation/test17699.d:1]: Add EXTRA_FILES for documentation purposes (and the gdc dejagnu testsuite)

**Comment by Geod24** [comment]
Should it, really ? I don't see it as particularly problematic, and the solution is to add one FS access for every module we open, which is IMO pretty bad.

**Comment by ibuclaw** [comment]
What happens if you `import pkg17699.datetime.foo;` when the package module is in this ambiguous structure? What happens if the module declarations are missing?

**Comment by dkorpel** [comment]
> Should it, really ? I don't see it as particularly problematic, 

See the bugzilla issue for rationale.

> and the solution is to add one FS access for every module we open, which is IMO pretty bad.

How much do you think it affects build times?

**Comment by dkorpel** [comment]
> What happens if you `import pkg17699.datetime.foo;` when the package module is in this ambiguous structure? 

The same as before.

> What happens if the module declarations are missing?

I don't think they affect anything here.

**Comment by Geod24** [comment]
> See the bugzilla issue for rationale.

Yeah that rationale is shaky. "If one version of Phobos is unzipped over another" isn't a great start, as it suffers from numerous other bugs as soon as the structure is touched.

> How much do you think it affects build times?

This instance, probably little, however it's the little things that matter. Any time we have to do an extra check / some extra work in the general case for a corner case, we should ask ourselves if it's the right thing to do. This case being, IMO, a textbook example.

**Comment by ibuclaw** [comment]
> > What happens if you `import pkg17699.datetime.foo;` when the package module is in this ambiguous structure? 
> 
> The same as before.
> 
> > What happens if the module declarations are missing?
> 
> I don't think they affect anything here.

I recall encountering a segfault when reducing something unrelated a few months back, didn't have time to synthesize. This file structure looked familiar though.

**Comment by dkorpel** [comment]
> This case being, IMO, a textbook example.

Textbook example of what? Do you have suggestions for a different implementation or should the issue just be closed as WONTFIX?

**Comment by ibuclaw** [comment]
> I recall encountering a segfault when reducing something unrelated a few months back, didn't have time to synthesize. This file structure looked familiar though.

https://issues.dlang.org/show_bug.cgi?id=23327

**Comment by schveiguy** [comment]
> Yeah that rationale is shaky. "If one version of Phobos is unzipped over another" isn't a great start, as it suffers from numerous other bugs as soon as the structure is touched.
> 

Rationale written in 2017, when a common mechanism to install your new compiler was to unzip it over the old one.

> and the solution is to add one FS access for every module we open

We are already opening multiple for a `package.d` form, plus opening one file per import directory to see if it's there.

Let's say you have a dub project with no dependencies. importing std.stdio stats 12 files before it eventually opens the right one (tested with strace):

```trace
stat("std/stdio.di", 0x7ffde5204920)    = -1 ENOENT (No such file or directory)
stat("std/stdio.d", 0x7ffde5204920)     = -1 ENOENT (No such file or directory)
stat("std/stdio.i", 0x7ffde5204920)     = -1 ENOENT (No such file or directory)
stat("std/stdio.c", 0x7ffde5204920)     = -1 ENOENT (No such file or directory)
stat("std/stdio", 0x7ffde5204920)       = -1 ENOENT (No such file or directory)
stat("source/std/stdio.di", 0x7ffde5204920) = -1 ENOENT (No such file or directory)
stat("source/std/stdio.d", 0x7ffde5204920) = -1 ENOENT (No such file or directory)
stat("source/std/stdio.i", 0x7ffde5204920) = -1 ENOENT (No such file or directory)
stat("source/std/stdio.c", 0x7ffde5204920) = -1 ENOENT (No such file or directory)
stat("source/std/stdio", 0x7ffde5204920) = -1 ENOENT (No such file or directory)
stat("/home/steves/.dvm/compilers/dmd-2.099.1/linux/bin/../../src/phobos/std/stdio.di", 0x7ffde5204920) = -1 ENOENT (No such file or directory)
stat("/home/steves/.dvm/compilers/dmd-2.099.1/linux/bin/../../src/phobos/std/stdio.d", {st_mode=S_IFREG|0664, st_size=173318, ...}) = 0
```

Add 5 more stats for each dependency.

Adding one more stat doesn't seem like it will make a huge difference in the performance.

File: compiler/test/fail_compilation/test17699.d (hunk 1 above)
Comment (dkorpel) [reply compiler/test/fail_compilation/test17699.d:1]: Done

**Comment by RazvanN7** [comment]
In this case I agree with @Geod24 . Even if the penalty is small compared to what is already being done, each little corner case check adds to the churn. On the hand, the benefit is almost non existent in my opinion, since the frequency of such a scenario is very small.

**Comment by schveiguy** [comment]
> Even if the penalty is small compared to what is already being done

Please measure. If this is going to add a few ns per import, I think it's worth avoiding causing a 1 hour puzzlement when it does happen.

Note, ImportC was worth adding N extra stats (one per import path), and nobody blinked. This is *one* extra stat, regardless of where the import is found.

**Comment by RazvanN7** [comment]
My point was more about adding ugliness to support a narrow use case, rather than performance. I don't think that the compiler should take care of every esoteric use case that arises.

That being said, I really don't care that much about this case. If others decide that this is worth it, I'm not going to block this addition in any way.

**Comment by schveiguy** [comment]
FYI, #14582 reduces the number of `stats` by over 90% for my project. After that PR is merged, an extra check is moot.

**Comment by thewilsonator** [comment]
Please retitle the _commit message_ "Fix bugzilla issue 17699: ..."
